sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala

```diff
@@ -379,6 +380,75 @@ class TypeCoercionSuite extends PlanTest {
     widenTest(ArrayType(IntegerType), StructType(Seq()), None)
   }
 
+  test("wider common type for decimal and array") {
+    def widenTest(t1: DataType, t2: DataType, tightestCommon: Option[DataType]) {
+      var found = TypeCoercion.findWiderTypeForTwo(t1, t2)
+      assert(found == tightestCommon,
+        s"Expected $tightestCommon as wider common type for $t1 and $t2, found $found")
+      // Test both directions to make sure the widening is symmetric.
+      found = TypeCoercion.findWiderTypeForTwo(t2, t1)
+      assert(found == tightestCommon,
+        s"Expected $tightestCommon as wider common type for $t2 and $t1, found $found")
+    }
+
+    def widenTestWithoutStringPromotion(
+        t1: DataType,
+        t2: DataType,
+        tightestCommon: Option[DataType]) {
+      var found = TypeCoercion.findWiderTypeWithoutStringPromotionForTwo(t1, t2)
+      assert(found == tightestCommon,
+        s"Expected $tightestCommon as wider common type for $t1 and $t2, found $found")
+      // Test both directions to make sure the widening is symmetric.
+      found = TypeCoercion.findWiderTypeWithoutStringPromotionForTwo(t2, t1)
+      assert(found == tightestCommon,
+        s"Expected $tightestCommon as wider common type for $t2 and $t1, found $found")
+    }
+
+    // Decimal
+    widenTest(DecimalType(2, 1), DecimalType(3, 2), Some(DecimalType(3, 2)))
+    widenTest(DecimalType(2, 1), DoubleType, Some(DoubleType))
+    widenTest(DecimalType(2, 1), IntegerType, Some(DecimalType(11, 1)))
+    widenTest(DoubleType, DecimalType(2, 1), Some(DoubleType))
+    widenTest(IntegerType, DecimalType(2, 1), Some(DecimalType(11, 1)))
+    widenTest(LongType, DecimalType(2, 1), Some(DecimalType(21, 1)))
+
+    // ArrayType
+    widenTest(ArrayType(NullType), ArrayType(ByteType), Some(ArrayType(ByteType)))
```

Comment:
These tests are too similar. we just need one test case for numeric, one for string, one for nested array.